DampedCosineCorrelationFunction.cpp

```diff
@@ -0,0 +1,122 @@
+//
+//    CLASSIFICATION:    Unclassified
+//
+//    DESCRIPTION:
+//
+//    This class is used to compute the correlation coefficient for a given
+//    delta time.
+//
+//    LIMITATIONS:       None
+//
+//    SOFTWARE HISTORY:
+//     Date          Author   Comment
+//     -----------   ------   -------
+//     22-Nov-2021   JPK      Copied from DampedCosineCorrelationModel
+//
+//    NOTES:
+//     Refer to DampedCosineCorrelationFunction.h for more information.
+//
+//#############################################################################
+
+#define CSM_LIBRARY
+#include "DampedCosineCorrelationFunction.h"
+#include "Error.h"
+#include <cmath>
+
+namespace csm {
+     
+DampedCosineCorrelationFunction::DampedCosineCorrelationFunction()
+   :
+      SPDCorrelationFunction("DampedCosineCorrelation"),
+      theCorrParams         ()
+{}
+
+DampedCosineCorrelationFunction::DampedCosineCorrelationFunction(double argA,
+                                                                 double argT,
+                                                                 double argP)
+   :
+      SPDCorrelationFunction("DampedCosineCorrelation"),
+      theCorrParams         ()
+{
+   setCorrelationParameters(argA,argT,argP);
+}
+      
+DampedCosineCorrelationFunction::
+DampedCosineCorrelationFunction(const Parameters& params)
+   :
+      SPDCorrelationFunction("DampedCosineCorrelation"),
+      theCorrParams         ()
+{
+   setCorrelationParameters(params);
+}
+
+DampedCosineCorrelationFunction::~DampedCosineCorrelationFunction()
+{}
+
+void DampedCosineCorrelationFunction::setCorrelationParameters(double argA,
+                                                               double argT,
+                                                               double argP)
+{
+   setCorrelationParameters(Parameters(argA,argT,argP));
+}
+
+void
+DampedCosineCorrelationFunction::setCorrelationParameters(const Parameters& params)
+{
+   static const char* const MODULE =
+      "csm::DampedCosineCorrelationFunction::setCorrelationParameters";
+
+   // make sure the values of each correlation function parameter fall within
+   // acceptable ranges
+   if ((params.A < 1.0e-6) || (params.A > 1.0))
+   {
+      throw Error(Error::BOUNDS,
+                  "Correlation parameter A must be in the range [1.0e-6, 1].",
+                  MODULE);
+   }
+
+   if (params.T < 1.0e-6)
+   {
+      throw Error(Error::BOUNDS,
+                  "Correlation parameter T must be >= 1.0e-6.",
+                  MODULE);
+   }
+   
+   if (params.P < 1.0e-6)
+   {
+      throw Error(Error::BOUNDS,
+                  "Correlation parameter P must be >= 1.0e-6.",
+                  MODULE);
+   }
+   // store the correlation parameter values
+   theCorrParams = params;
+}
+
+double
+DampedCosineCorrelationFunction::getCorrelationCoefficient(double deltaTime) const
+{
+   // compute the value of the correlation coefficient
+   constexpr double pi = 3.14159265358979323846;
```

Comment:
You may have to add an include for `<cmath>` and add `-D_USE_MATH_DEFINES` in `Makefile.win32`

Reply:
I'm not convinced this is the right thing to do.  While I realize many compilers do include M_PI in <cmath>, it isn't actually part of the language.  Adding that compile flag in Makefile.win32 may make it easy for someone to build the csm API library standalaone, but software toolkits (such as MSP) build the csm API library themselves in their own way... Making the change in Makefile.win32 won't have any impact on them...       PI is not a defined constant for C++ until C++20 when it will be available as a mathematical constant... and unfortunately years away from C++20.

Reply:
You could always 
#ifndef M_PI
#define M_PI 3.14...;
#endif

Reply:
@johnpdpkarp Modifying `Makefile.win32` would be OK because this constant is only used in the `.cpp` file, not a header file.  Or you could do like @RubeRad  suggests, but I think getting `M_PI` from the compiler is probably better.

Reply:
Better to have a csm constant for PI I think.